### PR TITLE
add overrie with list for each function with vararg parameters

### LINKED ISSUE #737: add override with List for each function with vararg parameters
I often write source codes like this for many function invocations. I think it's not good for performance and it's not good readable code.
```
type.parameterizedBy(*typeArguments.toTypedArray())
```
if there is override with List, I can write like this.
```
type.parameterizedBy(typeArguments)
```

### PR DESCRIPTION
**Pull request by anatawa12**
Resolves #737
Some override using List to less allocation.

### CODE REVIEW

File: kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt [hunk 1]
```diff
@@ -378,26 +378,52 @@ class FunSpec private constructor(
       parameters += parameterSpec
     }
 
+    @JvmName("callThisConstructorWithStrings")
```
Comment (Egorand) [inline kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt]: I really don't like the name and I can't see a better alternative. I think we should just drop the overloads that require changing method names.

File: kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt [hunk 2]
```diff
@@ -378,26 +378,52 @@ class FunSpec private constructor(
...
+    fun callThisConstructor(args: Iterable<CodeBlock>) = apply {
+      callConstructor("this", args.toList())
+    }
+
     fun callThisConstructor(vararg args: String) = apply {
-      callConstructor("this", *args.map { CodeBlock.of(it) }.toTypedArray())
+      callConstructor("this", args.map { CodeBlock.of(it) })
     }
 
     fun callThisConstructor(vararg args: CodeBlock = emptyArray()) = apply {
```
Comment (Egorand) [inline kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt]: It's interesting that we use an `emptyArray()` default arg here but not in the `vararg args: String` overload. This is also missing a `@JvmOverloads`. Out of scope of this PR, so no need to change.

File: kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt [hunk 3]
```diff
@@ -378,26 +378,52 @@ class FunSpec private constructor(
...
-      callConstructor("this", *args)
+      callConstructor("this", args.toList())
+    }
+
+    @JvmName("callSuperConstructorWithStrings")
```
Comment (Egorand) [inline kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt]: Same here, let's remove this overload.

File: kotlinpoet/src/main/java/com/squareup/kotlinpoet/jvm/JvmAnnotations.kt [hunk 4]
```diff
@@ -69,12 +69,24 @@ fun FunSpec.Builder.jvmName(name: String) = apply {
...
 
 fun FunSpec.Builder.throws(vararg exceptionClasses: Type) =
-    throws(*(exceptionClasses.map(Type::asTypeName).toTypedArray()))
+    throws(exceptionClasses.toList())
 
-fun FunSpec.Builder.throws(vararg exceptionClasses: TypeName) = addAnnotation(
+fun FunSpec.Builder.throws(vararg exceptionClasses: TypeName) =
+    throws(exceptionClasses.toList())
+
+@JvmName("throwsKClasses")
```
Comment (Egorand) [inline kotlinpoet/src/main/java/com/squareup/kotlinpoet/jvm/JvmAnnotations.kt]: Let's remove all these cause they require name changes.

File: kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt (hunk 2 above)
Comment (anatawa12) [reply kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt]: should I keep `private fun callConstructor(constructor: String, vararg args: CodeBlock)`? I think it makes duplicated code with `private fun callConstructor(constructor: String, vararg args: CodeBlock)` and `private fun callConstructor(constructor: String, args: List<CodeBlock>)`

File: kotlinpoet/src/main/java/com/squareup/kotlinpoet/jvm/JvmAnnotations.kt (hunk 4 above)
Comment (anatawa12) [reply kotlinpoet/src/main/java/com/squareup/kotlinpoet/jvm/JvmAnnotations.kt]: Fixed in 4f12e7d. could you check?

File: kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt (hunk 3 above)
Comment (anatawa12) [reply kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt]: Fixed in 4f12e7d. could you check?

File: kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt (hunk 2 above)
Comment (Egorand) [reply kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt]: No, I'm fine with the change you made. The arguments are ultimately kept as a list, so it doesn't matter at which point we convert a `vararg` to a list.

**Comment by anatawa12** [comment]
Thank you for merging my PR!

**Pull request merged**